components/infra-proxy-service/server/clients.go

```diff
@@ -2,31 +2,113 @@ package server
 
 import (
 	"context"
-	"sort"
+	"fmt"
 
 	chef "github.com/go-chef/chef"
 
 	"github.com/chef/automate/api/interservice/infra_proxy/request"
 	"github.com/chef/automate/api/interservice/infra_proxy/response"
 )
 
+// Client represents api client struct.
+type Client struct {
+	Name       string `json:"name"`
+	ClientName string `json:"clientname"`
+	OrgName    string `json:"orgname"`
+}
+
+// ClientResult list result from Search API
+type ClientResult struct {
+	Total int       `json:"total"`
+	Start int       `json:"start"`
+	Rows  []*Client `json:"rows"`
+}
+
 // GetClients get clients list
 func (s *Server) GetClients(ctx context.Context, req *request.Clients) (*response.Clients, error) {
 	c, err := s.createClient(ctx, req.OrgId, req.ServerId)
 	if err != nil {
 		return nil, err
 	}
 
-	clients, err := c.client.Clients.List()
+	res, err := c.SearchClients(req.SearchQuery)
 	if err != nil {
 		return nil, ParseAPIError(err)
 	}
 
 	return &response.Clients{
-		Clients: fromAPIToListClients(clients),
+		Clients: fromAPIToListClients(res.Rows),
+		Start:   int32(res.Start),
+		Total:   int32(res.Total),
 	}, nil
 }
 
+// SearchClients gets client list from Chef Infra Server search API.
+func (c *ChefClient) SearchClients(searchQuery *request.SearchQuery) (ClientResult, error) {
+	var result ClientResult
+	var searchAll bool
+	inc := 1000
+	var query chef.SearchQuery
+
+	if searchQuery == nil || searchQuery.Q == "" {
+		searchAll = true
+		query = chef.SearchQuery{
+			Index: "client",
+			Query: "*:*",
+			Start: 0,
+			Rows:  inc,
+		}
+	} else {
+		perPage := int(searchQuery.GetRows())
+		if perPage == 0 {
+			perPage = 1000
+		}
+
+		query = chef.SearchQuery{
+			Index: "client",
+			Query: searchQuery.GetQ(),
+			Start: int(searchQuery.GetStart()),
+			Rows:  perPage,
+		}
+	}
+
+	fullURL := fmt.Sprintf("search/%s", query)
+	newReq, err := c.client.NewRequest("GET", fullURL, nil)
+
+	if err != nil {
+		return result, ParseAPIError(err)
+	}
+
+	res, err := c.client.Do(newReq, &result)
+	if err != nil {
+		return result, ParseAPIError(err)
+	}
+
+	defer res.Body.Close() // nolint: errcheck
+
+	if searchAll {
+		var searchResult ClientResult
+		start := result.Start
+		// the total rows available for this query across all pages
+		total := result.Total
+		for start+inc <= total {
+			query.Start = query.Start + inc
+			fullURL = fmt.Sprintf("search/%s", query)
+
+			res1, err := c.client.Do(newReq, &searchResult)
+			if err != nil {
+				return result, ParseAPIError(err)
+			}
+
+			defer res1.Body.Close() // nolint: errcheck
```

Comment:
I think these defers are going to get stacked until the function returns. This is probably not good style, but can have some real consequences, such as this code wouldn't be able to reuse any of the http connections while in the loop

Reply:
Thanks for information @jaym , will rectify this

Reply:
Remove search all looping entirely, not using `res1.Body.Close()`.